### ISSUE TITLE
Workflow: a `compute` step type over a closed, versioned op registry

### ISSUE THREAD

**Issue by can-goodfire** [goodfire-ai/causalab#32]
## Context

Workflow v1 has exactly three step types — `protocol · select · plot`
(`causalab/protocol/workflow.py:67`, `docs/workflow_protocol.md` §2.1). Every
step therefore either runs an intervention protocol, ranks a metric table, or
draws one of two figures. There is no way to say *"transform this table (or
this tensor) into that one, deterministically"*.

That is the missing half of the analyses the refactor deleted. Fitting a
spline through class centroids, walking a geodesic, aggregating per-head
statistics, running a paired t-test — none of these touch a model, all of them
are pure functions of saved artifacts, and today each one is post-hoc analysis
outside the record. A pipeline that needs one is not expressible as a
document, which is exactly the property the workflow format exists to provide.

The obvious answer — a step that runs arbitrary Python — is the wrong one. It
would dissolve load-time validation (nothing to check), the canonical form
(nothing to canonicalize but a path), and determinism (§4's "a workflow run is
a pure function of the document"). The same argument the spec already makes
for the closed metric and plot vocabularies applies here.

## What exists today

| piece | file | state |
|---|---|---|
| step vocabulary | `causalab/protocol/workflow.py:67` `STEP_TYPES` | closed 3-tuple |
| step records | `causalab/protocol/workflow.py:91-125` | `ProtocolStep` / `SelectStep` / `PlotStep` |
| step parse | `causalab/protocol/workflow.py:204` `_parse_step` | dispatch on `type`, suggestions on unknown |
| dependency derivation | `causalab/protocol/workflow.py:504-580` | `after` + `from` + artifact refs + run-tree `file_path` loads |
| `from` restriction |  `causalab/protocol/workflow.py:522` | **`from` must name a `protocol` step** |
| ref restriction | `causalab/protocol/workflow.py:566-573` (rule 10) | an artifact ref must name a `select` step |
| canonical form | `causalab/protocol/workflow.py:877` `_canonicalize` | per-type branch; materializes defaults |
| execution | `causalab/workflow/runner.py:112-125` | one `isinstance` chain over the three types |

So the seam is small and well-defined: one new record, one parse branch, one
canonical branch, one runner branch — plus two restrictions that currently
assume only protocol steps produce tables.

## Proposal

A fourth step type whose `op` comes from a **closed-but-growable registry of
named, versioned, deterministic ops**:

```json
"fit": {
  "type": "compute", "op": "fit_pca@1",
  "inputs": {"acts": {"step": "harvest", "value": "acts.safetensors"}},
  "params": {"k": 8, "seed": 0},
  "outputs": {"basis": "basis.safetensors", "spectrum": "spectrum.parquet"}
}
```

| field | meaning |
|---|---|
| `op` | ✓ — `name@version` from the registry; unknown name or version is a load error with suggestions |
| `inputs` | ✓ — `{slot: {step, value}}`; each entry names a prior step's output, and **is** a derived dependency edge |
| `params` | – the op's declared parameters, validated against the op's own parameter schema at load |
| `outputs` | ✓ — `{slot: file_path}` under the step dir; the op declares its output slots, so a missing or extra slot is a load error |

**Why versioned names.** An op's numerics are part of the record: two runs of
the same document must agree. Pinning `fit_pca@1` means a behavioral change
ships as `fit_pca@2` and old documents keep digesting — and running — as
written. The version is part of the canonical form; the op's *implementation*
never is (the same rule that keeps backends out of protocol digests).

**Determinism is the registry's admission criterion.** An op must be a pure
function of its declared inputs and params. Anything stochastic takes an
explicit `seed` param and must be bit-stable across devices — the contract
`causalab/neural/pytorch_hooks/featurizers.py` already meets for seeded
featurizer init (CPU generator, then move). An op that cannot meet this does
not belong in the registry, and that refusal is the point of the closed set.

**Where the registry lives.** `causalab/compute/` — pure, torch-optional,
imported by the runner the same lazy way the reference backend is
(`causalab/protocol/cli.py:203`), so `validate`/`digest`/`explain` stay
free of a heavy dependency. Registration is a decorator carrying the op's
name, version, parameter schema, input slots and output slots; the load-time
checks read that record and never import the op body.

**Two restrictions to loosen, deliberately and narrowly:**

1. `from` on `select`/`plot` may name a `compute` step (`workflow.py:522`), so
   a computed table can be ranked and plotted. The group-by derivation
   (`runner.py:215`) reads sweep axes from the producing *document*; a compute
   step has no axes, so its tables aggregate as one group — state this in §2.3
   rather than leaving it to `run_axes.get(...)` returning `()`.
2. Rule 10's artifact-ref half stays as it is: only `select` steps emit values
   tables. A compute step's outputs are *files*, reached by the run-tree
   `file_path` overlay (`workflow.py:456` `_walk_run_tree_paths`) — the
   channel #29 built and the one the spec commits to for tensors. **A compute
   op is not a new tensor-passing mechanism**; it reuses the existing one.

**Seed ops (2–3, chosen to prove the shape, not to cover the analyses):**

| op | inputs → outputs | why this one |
|---|---|---|
| `fit_pca@1` | a saved read tensor → basis + spectrum | the smallest real op; `causalab/io/sklearn_pca.py` already exists as the reference implementation |
| `head_stats@1` | a metric table → a per-(layer, head) summary table | a table→table op, so both directions of the IO contract are exercised |
| `paired_ttest@1` | two metric tables → a statistics table | proves multi-input slots and a scalar-output table |

`geodesic_path@1` and the spline family are deliberately **not** seed ops:
they belong with the manifold hybrids that consume them, and adding them here
would pin an API before its consumer exists.

## Implementation plan

1. **`causalab/compute/registry.py`** — `ComputeOp` record (name, version,
   param schema, input slots, output slots, callable) + `register` decorator +
   `lookup(name_at_version)`; unknown name/version raises with `suggest(...)`,
   the same helper the other closed enums use.
2. **`causalab/compute/ops/`** — the seed ops, one module each, each with its
   own unit tests against a hand-computed oracle.
3. **`causalab/protocol/workflow.py`** — `"compute"` into `STEP_TYPES`; a
   `ComputeStep` record; a `_parse_step` branch validating `op`, `inputs`,
   `params`, `outputs` against the registry record; `inputs` entries added to
   `deps`/`data_deps` exactly where run-tree loads are (`:575`); `from` on
   select/plot accepts a compute producer (`:522`); a `_canonicalize` branch
   emitting `op`, sorted `inputs`, materialized `params` (defaults filled, the
   `_canon_train` treatment) and `outputs`.
4. **`causalab/workflow/runner.py`** — a `_run_compute_step` in the
   `:112` dispatch: resolve each input through the overlay
   (`OverlayArtifacts.resolve_path`, `:74`), call the op, write each declared
   output into the step dir, refuse if the op returned a slot it did not
   declare or omitted one it did.
5. **`docs/workflow_protocol.md`** — §2.1 step-type row; a new §2.4 for
   `compute` (renumbering `plot`/`save`); §3 sentence that `inputs` are
   dependency edges; checklist rules 1 and 4 extended to op names and slots;
   §6 row (op version in the canonical form); §11 note that the registry is
   growable by PR, not by document.

## Tests

- `tests/protocol/test_workflow.py` — a `compute` step parses; an unknown op
  name and an unknown *version* of a known op both refuse with a suggestion;
  a missing input slot, an undeclared extra slot, and a bad param type each
  refuse; the derived graph puts the compute step after its input's producer.
- `tests/protocol/test_canonical.py` — the op version and materialized params
  enter the canonical form; **every existing workflow digest is unchanged**
  (the new type adds no key to documents that do not use it) — an acceptance
  criterion, not a hope.
- `tests/neural/pytorch_hooks/test_workflow_run.py` — the capstone: a
  `protocol → compute → plot` workflow runs through the CLI on tiny-random,
  the compute step's outputs land in its step dir, and the plot reads the
  computed table.
- Per-op unit tests under `tests/compute/`, each against a hand-computed
  oracle — the same standard the metric kinds are held to.
- A test asserting `causalab validate`/`digest` on a compute workflow never
  imports the op body (the lazy-import discipline `docs/CODEBASE.md` states).

## Non-goals

- **No arbitrary Python step.** If an op is not worth registering and
  versioning, it is post-hoc analysis over the saved tables — the same trade
  §0 makes for metrics and plots.
- **No new tensor-passing mechanism.** Files in the run tree, selected by
  `entry` where a bundle holds several (#29), stay the only tensor channel.
- **No loops, conditionals, or dynamic op selection** — v1's acyclic finite
  step graph is unchanged.
- **No manifold ops.** `fit_spline`, `geodesic_path`, `hellinger_pca`,
  `path_scores` arrive with the workflows that consume them.

## Acceptance

- A `protocol → compute → plot` workflow passes the §5 checklist, digests
  stably across runs, and executes end-to-end through the CLI on tiny-random.
- Every existing workflow's digest is byte-identical.
- An unknown op, an unknown version, a missing/extra output slot and a bad
  param each fail at **load**, before any step runs.
- Each seed op has a unit test against a hand-computed oracle, and a documented
  determinism claim (with its seed param, where stochastic).
- `docs/workflow_protocol.md` describes the step type, its rules, and the
  registry's admission criterion.

Depends on #20 (the protocol refactor). Sibling of the manifold-hybrid work,
which is this issue's first real consumer.
